feat(RELEASE-1252): make run-file-updates task idempotent

jinqi7 opened the pull request:
The PR is to make the task idempotent by ensuring that a new MR isn't
 created unnecessarily when one already exists with the same content
in the seed and replacements of paths parameter.
And it added some logic to check the keys in the paths parameter. If the
 keys exist but their content does not require any replacement, return
 success.

jinqi7 commented:
/retest

mmalina commented:
@jinqi7 is this ready for review or not? If not, you can remove the reviewers and add them when ready.

johnbieren commented on internal/tasks/process-file-updates-task/tests/test-process-file-updates-replacements-idempotent.yaml:
```diff
@@ -0,0 +1,121 @@
+apiVersion: tekton.dev/v1
+kind: Pipeline
+metadata:
+  name: test-process-file-updates-replacements-mockidempotent
```
this should match the filename

johnbieren commented on internal/tasks/process-file-updates-task/tests/test-process-file-updates-replacements-idempotent.yaml:
```diff
@@ -0,0 +1,121 @@
+  name: test-process-file-updates-replacements-mockidempotent
+spec:
+  description: |
+    Run the process-file-updates task with replacements. The resulting
```
Can you update this?

johnbieren commented on internal/tasks/process-file-updates-task/tests/test-process-file-updates-replacements-idempotent.yaml:
```diff
@@ -0,0 +1,121 @@
+  
+              cd "$(params.tempDir)/mockidempotent"
+              commits=$(git log --oneline | wc -l)
+              echo "Test that only 1 commits are present"
```
Hmm if it was idempotent shouldn't there be 0 new ones?

jinqi7 replied:
The commit is for setting up the environment. There is no more commit created by the files updates since a mocked MR  is returned.

jinqi7 commented:
/retest

johnbieren commented on internal/tasks/process-file-updates-task/tests/test-process-file-updates-replacements-idempotent.yaml:
```diff
@@ -0,0 +1,121 @@
+      workspaces:
+        - name: pipeline
+          workspace: tests-workspace
+  finally:
```
This isn't necessary. We don't need to clean up dirs after a pipelineRun, only CRs

johnbieren commented on internal/tasks/process-file-updates-task/process-file-updates-task.yaml:
Can you bump the version string and add a README entry? Preferably the .Y

johnbieren reviewed (commented):
I will wait to review the actual task code for this PR to be marked ready

jinqi7 commented:
> @jinqi7 is this ready for review or not? If not, you can remove the reviewers and add them when ready.
sorry for annoying

jinqi7 commented:
/retest

jinqi7 commented:
/retest

jinqi7 commented:
/retest

jinqi7 commented:
/retest

theflockers commented on internal/tasks/process-file-updates-task/process-file-updates-task.yaml:
```diff
@@ -243,7 +243,61 @@ spec:
+            fi
+            # if all the lines are matched, the MR is the same one
+            # it should exit 0 if the same MR exists 
+            if [[ "$foundMR" == true ]]; then
```
Could you add another echo here saying that an existent MR for the same change was found, so the pipeline user would know why his new MR was not created - or think this is a new MR.

jinqi7 replied:
Yes. Added.

mmalina commented on internal/tasks/process-file-updates-task/README.md:
Can you provide more details in the commit message and PR description? At least some basic description of how it's being made idempotent.

jinqi7 replied:
Added.

mmalina replied:
Thanks

mmalina commented on internal/tasks/process-file-updates-task/process-file-updates-task.yaml:
```diff
@@ -243,7 +243,61 @@ spec:
+        fi
+
+        openMRList=$(glab mr list -R "${UPSTREAM_REPO}" --search "Konflux release" |grep "^!"| cut -f1 | tr -d '!')
+        for mrNum in ${openMRList} ; do
```
It seems to me this whole logic is only for the use case that there is replacement to be done, but maybe one of the open MRs already contains it. That's definitely a valid use case.

But what if the change has already been merged? Then we will probably fail on line 235, saying there is nothing to change, right? But if you run the same release repeatedly, that use case shouldn't fail.

Now the question is how to handle this - we will need to separate these two cases:

- there is substitution to be done because the files already contain the new strings -> pass
- there is substitution to be done for some other reason, e.g. we didn't find the lines to be updated -> fail

jinqi7 replied:
> But what if the change has already been merged? Then we will probably fail on line 235, saying there is nothing to change, right? But if you run the same release repeatedly, that use case shouldn't fail.

 How to differentiate the MR is merged and the file doesn't need to be changed at all? Do we still need to check the merged PRs?

mmalina replied:
We might need to discuss this with the whole team to decide how to handle this. Originally when this was implemented, the idea was that if there is nothing to be updated, then there's definitely something wrong. But now that we want to take a release pipeline rerun into account, we shouldn't automatically fail on this. We will need to be more granular. If we can find the file to update and then we search for the line to update and find it, but the line already contains the target string, then that's still ok. We only fail if we cannot find the file to update or cannot match the line to update.

mmalina replied:
> Do we still need to check the merged PRs?

I guess that would be one way to handle it. But probably not necessary. It would be difficult to decide how far into the history we want to go.

johnbieren replied:
IMO, if the "goal string" aka the string you are trying to insert already exists in the proper file, exit 0. No need to check merged MRs. If an MR is open with the exact change already, say that (idempotent) and exit 0 (ideally printing the existing MR url). If the file to be changed doesn't exist, exit 1. I *think* this is in agreement with what Martin has been commenting

johnbieren replied:
But yes, the case where the change is already merged seems already addressed. It just has to be updated to no longer fail (the line 235 piece Martin mentioned)

jinqi7 replied:
If only the return value is changed from "Failed" to "Success" in the block of "line 235",  the test  case of "test-process-file-updates-replacements-error.yaml" fails.  I added a variable of "keyNotFound" to check if the key is correct.  And if the file is not valid, it's handled in line 157. Is that ok?

mmalina replied:
I am reading through what the script does and I am not sure now. Until now, if a key is not found, that replacement will be skipped, but at the end we check that at least one replacement happened, otherwise we fail.

So yes, I think what you say sounds correct - we should fail if we cannot find the target file or the target key. But if those are found and there is no diff once the replacement is applied, it's still ok.

But I noticed there is also this `seed` option. It seems that one should be ok - if you run it again, it will just override the file with the seed string again, so there will be no diff. Previously this would fail on "no diff", but as we said, we will no longer fail on "no diff".

jinqi7 replied:
> Until now, if a key is not found, that replacement will be skipped, but at the end we check that at least one replacement happened, otherwise we fail.

I didn't change the logic of seed/replacement before, just add/set the "keyNotFound" variable.  If there is no replacement and all the keys can be found, it return true because the MR may be merged. If some key can't be found, it fails.

mmalina replied:
I added a new comment here: https://github.com/konflux-ci/release-service-catalog/pull/727/files#r1906905448

Closing this thread.

mmalina commented on internal/tasks/process-file-updates-task/process-file-updates-task.yaml:
```diff
@@ -243,7 +243,61 @@ spec:
+
+        if [[ -s "${TEMP}"/tempMRFile.diff ]]; then
+            # replacements exist
+            awk '/^@@/ {match($0, /@@ ([^@]+) @@/, arr); print arr[1]}' "${TEMP}"/tempMRFile.diff |\
```
What does this do? Can you add a comment explaining what it will print?

jinqi7 replied:
It's to deal with a line like "@@ -4,7 +4,7 @@ $schema: /app-interface/app-interface-settings-1.yml".  Usually it should be "@@ -4,7 +4,7 @@" in a line and " $schema: /app-interface/app-interface-settings-1.yml" in a seperated line.

johnbieren replied:
I still don't get it :joy: but can you make that comment in the code? This will be lost once this is merged

jinqi7 replied:
OK. Added a comment.

johnbieren commented on internal/tasks/process-file-updates-task/process-file-updates-task.yaml:
```diff
@@ -236,14 +238,75 @@ spec:
+
+            if [[ ! -s "${TEMP}"/lineFile ]]; then
+                # only seed files  
+                foundMR=false
```
Is this needed since it is set on line 264?

jinqi7 replied:
Removed. Thanks

jinqi7 commented:
/retest

jinqi7 commented:
/retest

jinqi7 committed:
feat(RELEASE-1252): make run-file-updates task idempotent

Signed-off-by: Jing Qi <jinqi@redhat.com>

The PR is to make the run-file-updates task idempotent by ensuring that
 a new MR isn't created unnecessarily when one already exists with the
same content in the value of `seed` and `replacements` fields in `paths`
 parameter.
And it adds some logic to check all of the `key` in the `replacements`
 parameter. If all of the `key` exist but the content of the file does
 not require any update, return success.

jinqi7 commented:
/retest

johnbieren reviewed (commented):
lgtm, but I'll let others (with open threads) approve

mmalina commented on tasks/internal/process-file-updates-task/process-file-updates-task.yaml:
```diff
@@ -236,14 +238,66 @@ spec:
+            if [[ "$keyNotFound" == true ]]; then
+                echo -n "Failed" |tee "$(results.fileUpdatesState.path)"
+            else
+                echo -n "Success" |tee "$(results.fileUpdatesState.path)"
```
This is ok, but above we would still save an error string to `$(results.fileUpdatesInfo.path)`. We should probably only save an error if we decide we want to fail, otherwise we should save some string saying there was nothing to change?

But another thing: This whole thing is inside `if [ "${REPLACEMENTS_PERFORMED}" == 0 ]`. Notice that `REPLACEMENTS_PERFORMED` is initially set to empty string. And then if there are some replacements, it's initialized to 0. I think the idea was that if there are no replacements at all, but there are seeds, you will still continue (and not exit here). So this use case needs to be handled as well - maybe add another `if` after `if [ "${REPLACEMENTS_PERFORMED}" == 0 ]` doing: If there is no diff, then it's ok and exit.

jinqi7 replied:
> This is ok, but above we would still save an error string to `$(results.fileUpdatesInfo.path)`. We should probably only save an error if we decide we want to fail, otherwise we should save some string saying there was nothing to change?

OK. I can remove the error message when it fails.

> But another thing: This whole thing is inside `if [ "${REPLACEMENTS_PERFORMED}" == 0 ]`. Notice that `REPLACEMENTS_PERFORMED` is initially set to empty string. And then if there are some replacements, it's initialized to 0. I think the idea was that if there are no replacements at all, but there are seeds, you will still continue (and not exit here). So this use case needs to be handled as well - maybe add another `if` after `if [ "${REPLACEMENTS_PERFORMED}" == 0 ]` doing: If there is no diff, then it's ok and exit. 

I think it will change the original logic, right? If "replacements" >0  and "${REPLACEMENTS_PERFORMED}" == 0 and $keyNotFound" == true, it need to be fail before.

mmalina replied:
> OK. I can remove the error message when it fails.

No, we do want an error message when it fails (when the key is not found). But we don't want an error message on a successful result (in this case if key was found).

> If "replacements" >0 and "${REPLACEMENTS_PERFORMED}" == 0 and $keyNotFound" == true, it need to be fail before.

What is `"replacements"`? You mean the number of replacements, i.e. `$REPLACEMENTS_LENGTH`?

That use case is already covered - as I said, `"${REPLACEMENTS_PERFORMED}" == 0` implies there was at least one replacement.

> I think it will change the original logic, right?

Yes, the whole point of this PR is to change the logic. Previously, if after applying the replacements and seeds there is no diff, we would fail. But now, if there is no diff, it's still ok and we just stop and succeed. And the use case when there is no replacement, but there is a seed and the diff is empty is currently not supported - the current code would fail on that and it shouldn't.

jinqi7 replied:
> Yes, the whole point of this PR is to change the logic. Previously, if after applying the replacements and seeds there is no diff, we would fail. But now, if there is no diff, it's still ok and we just stop and succeed. And the use case when there is no replacement, but there is a seed and the diff is empty is currently not supported - the current code would fail on that and it shouldn't.

With the original script runs for the first time, there is a seed and `"${REPLACEMENTS_PERFORMED}" == 0`  is empty,  the original script fails in that situation. So, it still fails in the current code. And you think it need to change, right?  I don't think we should let it pass when the replacements don't work at the first run.

mmalina replied:
> "${REPLACEMENTS_PERFORMED}" == 0 is empty

What do you mean by this? Empty means `""`. 0 is something else. So which one do you mean?

The original script will do this: If there is seed and no replacement (`"${REPLACEMENTS_PERFORMED}" == ""`) it's ok as long as there is a diff. Otherwise it fails. We need to pass even if there is no diff for this use case. I understand that in theory the first run should really generate some diff, but there is no easy way to find out if the lack of diff is because the seed was already created in a previous run, or if there is no diff for some other reason. So we have no option but to pass.

jinqi7 replied:
I mean "${REPLACEMENTS_PERFORMED}" == 0.   Only then it enters this block(  #238 ~#248).

mmalina replied:
But that's not the path I'm talking about. That path is already covered in the `if [ "${REPLACEMENTS_PERFORMED}" == 0 ]` block. I'm talking about a scenario when there is seed, but no replacement. In that case this `if` block will not apply, because `REPLACEMENTS_PERFORMED` will be `""`.

jinqi7 replied:
> there is no easy way to find out if the lack of diff is because the seed was already created in a previous run, or if there is no diff for some other reason. So we have no option but to pass.

I think it need to fail  when there is "replacements", but it doesn't do anything either the first time or the second time or more. The seeds should have been done if it's run for the second time.

mmalina replied:
Let me repeat it again: I am talking about the scenario when there is seed, but no replacements.

But since you brought up the scenario when there are replacements: I think we already agreed before that we only fail if we find a "key not found" error, otherwise we pass even if nothing was changed. That's what your current code does. The only comment I had was that if we pass/succeed, we should save an error message.

> it doesn't do anything either the first time or the second time or more

You don't know if the current run is the first time or second time, so that's why we agreed that we will just pass - there is no way for us to know if the change already happened, or if there is nothing to be changed to begin with.

jinqi7 replied:
> But that's not the path I'm talking about. That path is already covered in the `if [ "${REPLACEMENTS_PERFORMED}" == 0 ]` block. I'm talking about a scenario when there is seed, but no replacement. In that case this `if` block will not apply, because `REPLACEMENTS_PERFORMED` will be `""`.

If only seeds is set,  it can pass currently. See the test case - test-process-file-updates-seed.yaml

mmalina replied:
> If only seeds is set, it can pass currently.

I don't think it will and that's the main point of this thread.

Here's what will happen: The seeds are processed, but won't do any change (maybe because it is a second run). The `if [ "${REPLACEMENTS_PERFORMED}" == 0 ]` will not apply here, so the script will carry on. 

There there is this new block that you added for searching through open MRs - this shouldn't make any difference here, because our current change has no diff, so it will not match any open MR. And now it will try to create a MR with zero diff. OK, so it will not fail explicitly, but I don't think it will work. We shouldn't try to create a MR with an empty diff - that just won't work.

mmalina replied:
Probably the `git_commit_and_push` will fail, if it does a simple `git commit` without the `--allow-empty` flag.

jinqi7 replied:
I really don't know what need to be changed. If it's not anything changed by this PR,  can you file another task for it?

jinqi7 replied:
> > If only seeds is set, it can pass currently.
> 
> I don't think it will and that's the main point of this thread.
> 
> Here's what will happen: The seeds are processed, but won't do any change (maybe because it is a second run). The `if [ "${REPLACEMENTS_PERFORMED}" == 0 ]` will not apply here, so the script will carry on.
> 
> There there is this new block that you added for searching through open MRs - this shouldn't make any difference here, because our current change has no diff, so it will not match any open MR. And now it will try to create a MR with zero diff. OK, so it will not fail explicitly, but I don't think it will work. We shouldn't try to create a MR with an empty diff - that just won't work.

Has it been covered by this test case - test-process-file-updates-seed.yaml?

jinqi7 replied:
> Here's what will happen: The seeds are processed, but won't do any change (maybe because it is a second run). The `if [ "${REPLACEMENTS_PERFORMED}" == 0 ]` will not apply here, so the script will carry on.

If it's the second run, a MR should  exist. If it's the first run, see the test case test-process-file-updates-seed.yaml.

mmalina replied:
> Has it been covered by this test case - test-process-file-updates-seed.yaml?

You haven't updated that test, so I can't see how it would be covered there. It's a new functionality that should be part of this PR. Because previously we didn't support zero diff runs - previously those were expected to fail, but not anymore.

mmalina replied:
> If it's the second run, a MR should exist.

That is incorrect. We're talking about not having a diff here. So in that case the MR is likely already merged. And you iterate over open MR, not merged MRs. But even if the MR was in your list, again, we have zero diff and you're comparing the diff to the MR diffs, so you wouldn't find anything.

jinqi7 replied:
If we still consider the MRs merged,  we need to use "git diff --cached" to know if an empty file is added. Added this check in the PR.

mmalina replied:
We're moving in the right direction - the new check is what I expected. I'm just not sure about the `--cached`, let's discuss it. I noticed that when we add the seed files, we do `git commit add` for the files, so yes, then they will be in index and will show via `git diff --cached`. For the other file updates there is no `git add`, so those other changes would not show up in `git diff --cached`. So far so good. But what do you mean when you say `to know if an empty file is added`? What do you mean by this? Can you specify the use case? I think the current approach will work, I just don't know what you mean by it.

If the file is already there and the seed does nothing, the `git add` has no effect. So both `git diff` and `git diff --cached` will be empty. But if the file was added, only `git diff --cached` will show it.

jinqi7 replied:
If there is seed and the file is already there,  #136~#144  adds the file again ( it does not fail). But `git diff --cached` return nothing ( It's the situation of `an empty file is added` ). If there is seed and the file is not there, #136~#144  adds the file. `git diff --cached` will show it.

mmalina replied:
Yeah, I just think it's confusing to say "an empty file is added" :) It's not empty. But it was already there, so nothing happens. OK, all good now. I think it should work.

jinqi7 commented:
/retest

jinqi7 commented:
/retest

jinqi7 commented:
/retest

jinqi7 commented:
/retest

jinqi7 commented:
/retest

jinqi7 commented:
/retest

jinqi7 commented:
/retest

jinqi7 commented:
/retest

jinqi7 commented:
/retest
